File: JAVA/ConstructBinaryTreefromInorderandPostorderTraversal.java
class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;

	TreeNode(int x) {
		val = x;
	}
}
class Solution {
	public TreeNode buildTree(int[] inorder, int[] postorder) {
        // Start typing your Java solution below
        // DO NOT write main() function
    	if(inorder == null || inorder.length != postorder.length)
    		return null;
        return build(inorder, 0, inorder.length-1, postorder, 0, postorder.length-1);
    }

	private TreeNode build(int[] inorder, int start1, int end1, int[] postorder,
			int start2, int end2) {
		if(start1 > end1 || start2 > end2)
			return null;
		TreeNode root = new TreeNode(postorder[end2]);
		int index = -1;
		for(int i = start1; i <= end1; i++)
		{
			if(inorder[i] == root.val)
			{
				index = i;
				break;
			}
		}
		root.left = build(inorder, start1, index-1, postorder, start2, start2+(index-start1)-1);
		root.right = build(inorder, index+1, end1, postorder, start2+(index-start1), end2-1);
		return root;
	}
	
}

public class ConstructBinaryTreefromInorderandPostorderTraversal {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}

File: C++/ConstructBinaryTreefromInorderandPostorderTraversal.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
using namespace std;
struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	int find(vector<int> list, int s, int e, int target)
	{
		for(int i = s; i <= e; i++)
		{
			if(list[i] == target)
				return i;
		}
		return -1;
	}
	TreeNode* build(vector<int>& inorder, int s1, int e1, vector<int>&postorder, int s2, int e2)
	{
		if(s1 > e1)
			return NULL;
		TreeNode* root = new TreeNode(postorder[e2]);
		int index = find(inorder, s1, e1, root->val);
		root->left = build(inorder, s1, index-1, postorder, s2, s2 + (index-s1)-1);
		root->right = build(inorder, index+1, e1, postorder, s2 + (index-s1), e2-1);
		return root;
	}
	TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder) {
		// Start typing your C/C++ solution below
		// DO NOT write int main() function
		int len = inorder.size();
		return build(inorder, 0, len-1, postorder, 0, len-1);
	}
};